src/hotspot/share/opto/parse3.cpp

```diff
@@ -77,10 +77,21 @@ void Parse::do_field_access(bool is_get, bool is_field) {
   Node* obj;
   if (is_field) {
     int obj_depth = is_get ? 0 : field->type()->size();
-    obj = null_check(peek(obj_depth));
+    Node* original_obj = peek(obj_depth);
+    obj = null_check(original_obj);
     // Compile-time detect of null-exception?
     if (stopped())  return;
 
+    if (DoPartialEscapeAnalysis) {
+      PEAState& as = jvms()->alloc_state();
+      PartialEscapeAnalysis *pea = PEA();
+      VirtualState* vs = as.as_virtual(pea, original_obj);
+      // The null-checked object node aliases to the same object as the original node.
+      if (vs != nullptr) {
+        pea->add_alias(pea->is_alias(original_obj), obj);
```

Comment:
so you want to say that castPP is same as original obj? 
Is it possible that we consider disjoint set over hashmap? 
it's faster to do union and changing 'root' in disjoint set.

Reply:
Yes, I think its possible. I think we can revisit our data structures in separate work.